core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/PartialHealthCheckStrategy.java

```diff
@@ -0,0 +1,202 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint.healthcheck;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.ThreadLocalRandom;
+
+import org.jctools.maps.NonBlockingHashSet;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.client.Endpoint;
+
+/**
+ * A strategy to check part of candidates by health.
+ */
+final class PartialHealthCheckStrategy implements HealthCheckStrategy {
+
+    /**
+     * Creates a new builder.
+     */
+    public static PartialHealthCheckStrategyBuilder builder() {
+        return new PartialHealthCheckStrategyBuilder();
+    }
+
+    private final Set<Endpoint> selectedCandidates;
+    private final Set<Endpoint> unhealthyCandidates;
+    private final TargetCount max;
+    private Set<Endpoint> candidates;
+
+    /**
+     * Creates a new instance.
+     */
+    PartialHealthCheckStrategy(TargetCount max) {
+        this.max = requireNonNull(max, "max");
+        selectedCandidates = new HashSet<>();
+        unhealthyCandidates = new NonBlockingHashSet<>();
+        candidates = ImmutableSet.of();
+    }
+
+    @Override
+    public void updateCandidates(List<Endpoint> candidates) {
+        requireNonNull(candidates, "candidates");
+
+        synchronized (selectedCandidates) {
+            this.candidates = ImmutableSet.copyOf(candidates);
+
+            final Set<Endpoint> removedCandidates = new HashSet<>();
+            for (Endpoint selectedCandidate : selectedCandidates) {
+                if (this.candidates.contains(selectedCandidate)) {
+                    continue;
+                }
+
+                removedCandidates.add(selectedCandidate);
+            }
+
+            removeAndSelectNewCandidates(removedCandidates);
+        }
+    }
+
+    @Override
+    public List<Endpoint> getCandidates() {
+        synchronized (selectedCandidates) {
+            return ImmutableList.copyOf(selectedCandidates);
+        }
+    }
+
+    @Override
+    public boolean updateHealth(Endpoint endpoint, double health) {
+        final double unhealthyScore = 0;
+
+        requireNonNull(endpoint, "endpoint");
+
+        if (!candidates.contains(endpoint)) {
+            unhealthyCandidates.remove(endpoint);
+            return true;
```

Comment:
Shouldn't we return `false` here because `selectedCandidates` didn't change?

Reply:
fixed

Reply:
I think the code below is a situation where we have to order `HealthCheckEndpointGroup` to update candidates. 
```java
 if (!candidates.contains(endpoint)) {
    unhealthyCandidates.remove(endpoint);
    return true;
}
```
The strategy manages candidates. If the strategy doesn't have an endpoint, it means that the  states between `HealthCheckEndpointGroup` and `PartialHealthCheckStrategy` is different.